### PR TITLE
Introduce the `subset` class kind as an annotation

### PR DESCRIPTION
**Pull request by jcbrinfo**
Each time a `subset` annotation is found in a class definition, set the class kind to `subset`. Also, implement the inheritance and instantiation rules.

### CODE REVIEW

**Comment by jcbrinfo** [comment]
♻️ 
- [x] Add a missing test module.

File: src/model/model.nit [hunk 1]
```diff
@@ -2572,6 +2572,12 @@ end
 class MClassKind
 	redef var to_s
 
+	# Can a class of kind `self` define a membership predicate?
+	var can_customize_isa: Bool
+
+	# Can a class of kind `self` define a constructor?
+	var can_init: Bool
```
Comment (privat) [inline src/model/model.nit]: aren't `can_customize_isa` and `can_init` both exclusive?

File: src/model/model.nit [hunk 2]
```diff
@@ -2586,19 +2586,23 @@ class MClassKind
...
+			return other != subset_kind
 		else if self == extern_kind then
 			# only compatible with themselves
 			return self == other
-		else if other == enum_kind or other == extern_kind then
-			# abstract_kind and concrete_kind are incompatible
-			return false
+		else
+			# assert self == abstract_kind or self == concrete_kind
+			return other == abstract_kind or other == concrete_kind
```
Comment (privat) [inline src/model/model.nit]: please, do not change the form of the existing `if` in a same commit that change the beahvoir. This make the commit unreadable.
By the way I preferred the previous approach with less complexity in the `return` and a systematic approche in the `if`. I'm not sure what to are trying to achieve by changing this.

**Comment by jcbrinfo** [comment]
I think I said to you that there was no conflict, but you insisted that I prototype with annotations. So, I did as you said.

File: src/model/model.nit (hunk 1 above)
Comment (jcbrinfo) [inline src/model/model.nit]: For now, yes. But, they are completely different concerns in my opinion. For example, some languages (like X10) allow concrete classes to define a predicate.

File: src/model/model.nit (hunk 2 above)
Comment (jcbrinfo) [inline src/model/model.nit]: We now have subsets. So, blacklisting enums and externs is not enough. Also, using most of the `if`s as a switch-case over `self` makes the whole logic look clearer to me. It also removes a case. Furthermore, you already put a comparison with `other` in a `return` above.

**Comment by jcbrinfo** [comment]
Obsoleted by #2483.

**Pull request closed without merging**